### PR TITLE
Load group program times separately from main query

### PR DESCRIPTION
**Pull request by hugo-vrijswijk**

### CODE REVIEW

File: build.mjs [hunk 1]
```diff
@@ -2,17 +2,22 @@ import path from 'path';
...
 import fs from 'fs/promises';
 import process from 'process';
-import { build } from 'vite';
+import { build, loadConfigFromFile } from 'vite';
 import { minify } from 'terser';
 import humanFormat from 'human-format';
 import brotliSize from 'brotli-size';
-import config from './vite.config.mjs';
+
+const config = await loadConfigFromFile({ mode: 'production' });
```
Comment (rpiaggio) [inline build.mjs:10]: Why this change?

File: explore/src/main/scala/explore/ExplorePWA.scala [hunk 2]
```diff
@@ -138,8 +135,8 @@ object ExplorePWA {
     bc.messages
       .evalTap { me =>
         me.data.event match
-          case ExploreEvent.PWAReloadId      => IO.fromFuture(IO(updateSW(true).toFuture)) // .void
-          case ExploreEvent.ExploreUIReadyId => IO.fromFuture(IO(updateSW(()).toFuture))   // .void
+          case ExploreEvent.PWAReloadId      => IO.fromPromise(IO(updateSW(true))) // .void
```
Comment (rpiaggio) [inline explore/src/main/scala/explore/ExplorePWA.scala:138]: Nice

File: model/shared/src/main/scala/explore/model/GroupElement.scala [hunk 3]
```diff
@@ -54,23 +58,50 @@ object GroupElement:
...
+  def findParentGroupIds(
+    groups:    GroupList,
+    elementId: Either[Observation.Id, Group.Id]
+  ): List[Group.Id] = {
+    @tailrec
+    def go(
+      acc:     List[Group.Id],
+      current: Either[Observation.Id, Group.Id]
+    ): List[Group.Id] =
+      val parentGroup = groups.find(
```
Comment (rpiaggio) [inline model/shared/src/main/scala/explore/model/GroupElement.scala:73]: To find the parent node, are we scanning the whole list of groups on each iteration?

**Review by rpiaggio — APPROVED** [review]
LGTM

I hadn't taken a look at the way Groups are implemented until now. I'd like to see if we can retrofit it into an `IndexedKeyTree` so that:
- We have separate structure from data.
- All nodes have a link to its parent.
- We can have positional undo, even in the face of intermediate remote edits/insertions/deletions. (I actually haven't checked if the current structure supports this).

File: build.mjs (hunk 1 above)
Comment (hugo-vrijswijk) [reply build.mjs:10]: I moved the vite config to a typescript file, but that means you can't import it as plain js anymore

File: model/shared/src/main/scala/explore/model/GroupElement.scala (hunk 3 above)
Comment (hugo-vrijswijk) [reply model/shared/src/main/scala/explore/model/GroupElement.scala:73]: Yes, if it's not in a tree structure, I think this is the fastest/only way. Since I imagine groups won't be _that_ deep, it won't be too many iterations. `IndexedKeyTree` is already used, but I can have a look at lifting it higher

File: explore/src/main/scala/explore/cache/ProgramCache.scala [hunk 4]
```diff
@@ -213,6 +241,15 @@ object ProgramCache
               queryProgramTimes
         )
 
+        val groupTimeRangeUpdate: Pipe[
+          IO,
+          ProgramQueriesGQL.GroupEditSubscription.Data,
+          ProgramSummaries => ProgramSummaries
+        ] = keyedSwitchEvalMap(
+          _.groupEdit.value.map(_.id),
+          data => updateGroupTimeRange(props.programId)
```
Comment (toddburnside) [inline explore/src/main/scala/explore/cache/ProgramCache.scala:250]: Should this also flatTap to queryProgramTimes?

**Review by toddburnside — APPROVED** [review]
LGTM, with one question.

File: explore/src/main/scala/explore/cache/ProgramCache.scala (hunk 4 above)
Comment (hugo-vrijswijk) [reply explore/src/main/scala/explore/cache/ProgramCache.scala]: Yes, I've added that now
